rust/agents/relayer/src/msg/gelato_submitter.rs

```diff
@@ -1,98 +1,231 @@
+        }
+        Ok(())
+    }
 
-        // TODO(webbhorn): Scan pending queue for any newly-eligible
-        // ops and if encountered, spawn them in a root task.
-        // Remove them from pending queue if so.
+    fn make_forward_request_args(&self, msg: SubmitMessageArgs) -> Result<ForwardRequestArgs> {
+        let ivm_base_contract = BaseContract::from(ivm_abi.clone());
+        let call_data = ivm_base_contract.encode(
+            "process",
+            [
+                Token::Address(self.inbox_address),
+                Token::FixedBytes(msg.checkpoint.checkpoint.root.to_fixed_bytes().into()),
+                Token::Uint(msg.checkpoint.checkpoint.index.into()),
+                Token::Array(
+                    msg.checkpoint
+                        .signatures
+                        .iter()
+                        .map(|s| Token::Bytes(s.to_vec()))
+                        .collect(),
+                ),
+                Token::Bytes(msg.committed_message.message.to_vec()),
+                Token::FixedArray(
+                    (0..32)
+                        .map(|i| Token::FixedBytes(msg.proof.path[i].to_vec()))
+                        .collect(),
+                ),
+                Token::Uint(msg.leaf_index.into()),
+            ],
+        )?;
+        Ok(ForwardRequestArgs {
+            chain_id: abacus_domain_to_gelato_chain(self.inbox_domain)?,
+            target: self.ivm_address,
+            data: call_data,
+            fee_token: NATIVE_FEE_TOKEN_ADDRESS,
+            payment_type: PaymentType::AsyncGasTank,
+            max_fee: DEFAULT_MAX_FEE.into(), // Maximum fee that sponsor is willing to pay.
+            gas: DEFAULT_MAX_FEE.into(),     // Gas limit.
+            sponsor_chain_id: abacus_domain_to_gelato_chain(self.outbox_domain)?,
+            nonce: U256::zero(),
+            enforce_sponsor_nonce: false,
+            enforce_sponsor_nonce_ordering: false,
+            sponsor: self.sponsor_address,
+        })
+    }
+}
 
-        // TODO(webbhorn): Either wait for finality in the ForwardRequestOp
-        // logic, or follow the pattern from serial_submitter.rs
-        // of implementing a verification queue, where we will stash
-        // successfully submitted ops that have not yet reached finality.
-        // Only after reaching finality will we commit the new status to
-        // AbacusDB and drop those messages from the verification queue.
-        // In case of a destination chain re-org, they would need
-        // to go back to the wait queue.
+// TODO(webbhorn): Is there already somewhere actually canonical/authoritative to use instead
+// of duplicating this here?  Perhaps we can expand `macro_rules! domain_and_chain`?
+// Otherwise, try to keep this translation logic out of the gelato crate at least so that we
+// don't start introducing any Abacus concepts (like domain) into it.
+fn abacus_domain_to_gelato_chain(domain: u32) -> Result<Chain> {
+    Ok(match domain {
```

Comment:
You could move this to where we do the definition for the domain ids, that way duplication is at least closer together. As for a way to do it with a macro, not really possible because the names are not the same. Might also consider using the domain id enum instead of using u32s, that way it forces all of the domain IDs to get mapped.

Reply:
Hm ok yeah gonna keep thinking about this and maybe leave for now, would love to normalize some of this chain config we've got floating around (typescript infra config, rust json config, some stuff hardcoded...). Really the only reason we have a separate Gelato chain enum right now is so that we can get the address of the hardcoded verification contract addresses, which are required in the domain hash for the eip 712 signature, and afaict not available any other way. I'm reluctant to stuff that in abacus-core/srcj/chains.rs, esp when we're not 1:1 for support

Reply:
Hmm sorry not sure I have an opinion yet